src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,248 @@
+    {
+        $this->configuration = array_replace_recursive(self::DEFAULT_OPTIONS, $configuration);
+        $this->driverConnection = $driverConnection;
+    }
+
+    public function getConfiguration(): array
+    {
+        return $this->configuration;
+    }
+
+    public static function buildConfiguration($dsn, array $options = [])
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The given Doctrine DSN "%s" is invalid.', $dsn));
+        }
+
+        $components = parse_url($dsn);
+        $query = [];
+        if (isset($components['query'])) {
+            parse_str($components['query'], $query);
+        }
+
+        $configuration = [
+            'connection' => $components['host'],
+            'table_name' => $options['table_name'] ?? ($query['table_name'] ?? self::DEFAULT_OPTIONS['table_name']),
+            'queue_name' => $options['queue_name'] ?? ($query['queue_name'] ?? self::DEFAULT_OPTIONS['queue_name']),
+            'redeliver_timeout' => $options['redeliver_timeout'] ?? ($query['redeliver_timeout'] ?? self::DEFAULT_OPTIONS['redeliver_timeout']),
+            'loop_sleep' => $options['loop_sleep'] ?? ($query['loop_sleep'] ?? self::DEFAULT_OPTIONS['loop_sleep']),
+            'auto_setup' => $options['auto_setup'] ?? ($query['auto_setup'] ?? self::DEFAULT_OPTIONS['auto_setup']),
+        ];
+
+        // check for extra keys in options
+        $optionsExtraKeys = array_diff(array_keys($options), array_keys($configuration));
+        if (0 < \count($optionsExtraKeys)) {
+            throw new TransportException(sprintf('Unknown option found : [%s]. Allowed options are [%s]', implode(', ', $optionsExtraKeys), implode(', ', self::DEFAULT_OPTIONS)));
+        }
+
+        // check for extra keys in options
+        $queryExtraKeys = array_diff(array_keys($query), array_keys($configuration));
+        if (0 < \count($queryExtraKeys)) {
+            throw new TransportException(sprintf('Unknown option found in DSN: [%s]. Allowed options are [%s]', implode(', ', $queryExtraKeys), implode(', ', self::DEFAULT_OPTIONS)));
+        }
+
+        return $configuration;
+    }
+
+    /**
+     * @param int $delay The delay in milliseconds
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function send(string $body, array $headers, int $delay = 0): void
+    {
+        $now = (\DateTime::createFromFormat('U.u', microtime(true)));
+        $availableAt = (clone $now)->modify(sprintf('+%d seconds', $delay / 1000));
+
+        $sql = $this->driverConnection->createQueryBuilder()
+            ->insert($this->configuration['table_name'])
+            ->values([
+                'body' => ':body',
+                'headers' => ':headers',
+                'queue_name' => ':queue_name',
+                'created_at' => ':created_at',
+                'available_at' => ':available_at',
+            ]);
```

Comment:
We could change `$sql` to `$queryBuilder` and below, use `$queryBuilder->getSql()`. That would avoid relying on the `__toString()` method in `QueryBuilder`, which is nice, because if there was an error for some reason generating the SQL, it'll get buried because it's inside of `__toString()`.